File: Heap Sort/src/BinaryHeap.java
import java.util.Arrays;

/**
 * Created by shawon on 5/24/16.
 */
public class BinaryHeap {
    private int[] a;
    private int length;
    private int heapSize;

    public BinaryHeap() {}
    public BinaryHeap(int[] a) {
        this.a = a;
        length = a.length;
        heapSize = length;
    }

    //=======================

    public int[] getA() {
        return a;
    }

    public void setA(int[] a) {
        this.a = a;
    }

    public int getLength() {
        return length;
    }

    public void setLength(int length) {
        this.length = length;
    }

    public int getHeapSize() {
        return heapSize;
    }

    public void setHeapSize(int heapSize) {
        this.heapSize = heapSize;
    }

    //================
    public int getRootIndex() { return 0; }
    public int getLeftSubIndex(int i) { return ((2 * i) + 1); }
    public int getRightSubIndex(int i) { return ((2 * i) + 2); }
    public int getParentIndex(int i) { return (i / 2); }

    //=============================

    public void maxHeapify(int i) {
        int leftSubIndex = getLeftSubIndex(i);
        int rightSubIndex = getRightSubIndex(i);
        int largestElementIndex;

        if ((leftSubIndex < heapSize) && a[leftSubIndex] > a[i]) largestElementIndex = leftSubIndex;
        else largestElementIndex = i;

        if (rightSubIndex < heapSize && a[rightSubIndex] > a[largestElementIndex]) largestElementIndex = rightSubIndex;

        if(largestElementIndex != i) {
            // swap
            int temp = a[i];
            a[i] = a[largestElementIndex];
            a[largestElementIndex] = temp;

            // call max heapify again
            maxHeapify(largestElementIndex);
        }
    }

    public void buildMaxHeap() {
        for(int i = (length / 2) - 1; i >= 0; i--) {
            maxHeapify(i);
        }
    }

    public void minHeapify(int i) {
        int leftSubIndex = getLeftSubIndex(i);
        int rightSubIndex = getRightSubIndex(i);
        int smallestElementIndex;

        if ((leftSubIndex < heapSize) && a[leftSubIndex] < a[i]) smallestElementIndex = leftSubIndex;
        else smallestElementIndex = i;

        if (rightSubIndex < heapSize && a[rightSubIndex] < a[smallestElementIndex]) smallestElementIndex = rightSubIndex;

        if(smallestElementIndex != i) {
            // swap
            int temp = a[i];
            a[i] = a[smallestElementIndex];
            a[smallestElementIndex] = temp;

            // call min heapify again
            minHeapify(smallestElementIndex);
        }
    }

    public void buildMinHeap() {
        for(int i = (length / 2) - 1; i >= 0; i--) {
            minHeapify(i);
        }
    }

    //========================
    public void heapSortAscending() {
        buildMaxHeap();

        for(int i = length - 1; i >= 1; i--) {
            // swap
            int temp = a[0];
            a[0] = a[i];
            a[i] = temp;
            heapSize--;

            // call maxHeapify
            maxHeapify(0);
        }
    }

    public void heapSortDescending() {
        buildMinHeap();

        for(int i = length - 1; i >= 1; i--) {
            // swap
            int temp = a[0];
            a[0] = a[i];
            a[i] = temp;
            heapSize--;

            // call minHeapify
            minHeapify(0);
        }
    }

    //==================

    public void printBinaryHeap() {
        System.out.println("Binary Heap contains : " + Arrays.toString(a));
    }

}

File: Priority Queue C Sharp/BinaryHeap.cs
using System;

namespace ConsoleApplication
{
    public class BinaryHeap
    {
        private int[] _a;
        private int _length;
        private int _heapSize;

        //=====================

        public BinaryHeap() {}

        public BinaryHeap(int[] a)
        {
            _a = a;
            _length = a.Length;
            _heapSize = _length;
        }

        //======================

        public int[] A
        {
            get
            { return A; }
            set
            { value = _a; }
        }

        public int Length { get { return _length; } set { value = _length; } }
        public int HeapSize { get { return _heapSize; } set { value = _heapSize; } }

        //===============================

        public int GetRootIndex() { return 0; }
        public int GetLeftSubIndex(int i) { return ((2 * i) + 1); }
        public int GetRightSubIndex(int i) { return ((2 * i) + 2); }
        public int GetParentIndex(int i) { return (i / 2); }

        //==============================

        public void MaxHeapify(int i) {
          int leftSubIndex = GetLeftSubIndex(i);
          int rightSubIndex = GetRightSubIndex(i);
          int largestElementIndex;

          if ((leftSubIndex < _heapSize) && _a[leftSubIndex] > _a[i]) largestElementIndex = leftSubIndex;
          else largestElementIndex = i;

          if (rightSubIndex < _heapSize && _a[rightSubIndex] > _a[largestElementIndex]) largestElementIndex = rightSubIndex;

          if(largestElementIndex != i) {
              // swap
              int temp = _a[i];
              _a[i] = _a[largestElementIndex];
              _a[largestElementIndex] = temp;

              // call max heapify again
              MaxHeapify(largestElementIndex);
            }
        }

        public void BuildMaxHeap() {
          for(int i = (_length / 2) - 1; i >= 0; i--) {
              MaxHeapify(i);
            }
        }

        public void MinHeapify(int i) {
          int leftSubIndex = GetLeftSubIndex(i);
          int rightSubIndex = GetRightSubIndex(i);
          int smallestElementIndex;

          if ((leftSubIndex < _heapSize) && _a[leftSubIndex] < _a[i]) smallestElementIndex = leftSubIndex;
          else smallestElementIndex = i;

          if (rightSubIndex < _heapSize && _a[rightSubIndex] < _a[smallestElementIndex]) smallestElementIndex = rightSubIndex;

          if(smallestElementIndex != i) {
              // swap
              int temp = _a[i];
              _a[i] = _a[smallestElementIndex];
              _a[smallestElementIndex] = temp;

              // call min heapify again
              MinHeapify(smallestElementIndex);
            }
        }

        public void BuildMinHeap() {
          for(int i = (_length / 2) - 1; i >= 0; i--) {
              MinHeapify(i);
          }
        }

        public void PrintBinaryHeap()
        {
            Console.Write("Heap Data : [ ");
            foreach (var h in A)
            {
              Console.Write(h + ", ");
            }
            Console.WriteLine(" ]");
        }
    }
}
